src/darwin/Framework/CHIP/MTRCommissioningOperation.mm

```diff
@@ -0,0 +1,490 @@
+}
+
+- (void)_earlyFailCommissioning:(CHIP_ERROR)error
+{
+    // This handles failures before we have cleared the way for doing a commissioning at all, and in
+    // particular we might have an existing commissioning when we land here.  Don't mess with the
+    // "normal" metrics here, because we might stomp on the metrics for that ongoing commissioning.
+    auto * emptyMetrics = [[MTRMetrics alloc] initWithCapacity:0];
+    [self _dispatchCommissioningError:[MTRError errorForCHIPErrorCode:error] withMetrics:emptyMetrics];
+}
+
+- (void)_dispatchCommissioningCHIPError:(CHIP_ERROR)error
+{
+    MATTER_LOG_METRIC_END(kMetricDeviceCommissioning, error);
+    MTRMetrics * metrics = [[MTRMetricsCollector sharedInstance] metricSnapshotForCommissioning:YES];
+    [self _dispatchCommissioningError:[MTRError errorForCHIPErrorCode:error] withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error
+{
+    MATTER_LOG_METRIC_END(kMetricDeviceCommissioning, [MTRError errorToCHIPErrorCode:error]);
+    MTRMetrics * metrics = [[MTRMetricsCollector sharedInstance] metricSnapshotForCommissioning:YES];
+    [self _dispatchCommissioningError:error withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error withMetrics:(MTRMetrics *)metrics
+{
+    [self _dispatchCommissioningError:error forCommissioningID:_commissioningID withMetrics:metrics];
+}
+
+- (void)_dispatchCommissioningError:(NSError *)error forCommissioningID:(NSNumber *)commissioningID withMetrics:(MTRMetrics *)metrics
+{
+    MTRDeviceController_Concrete * strongController = _controller;
+
+    MTR_LOG("%@ Device commissioning failed with controller %@ metrics %@", self, strongController, metrics);
+
+    id<MTRCommissioningDelegate> strongDelegate = _delegate;
+    id<MTRCommissioningDelegate_Internal> strongInternalDelegate = [self _internalDelegate];
+
+    // Null out the delegate, so we don't notify based on any internal cleanups
+    // we happen to do.
+    _delegate = nil;
+
+    dispatch_async(_delegateQueue, ^{
+        if ([strongInternalDelegate respondsToSelector:@selector(commissioning:failedWithError:forDeviceID:metrics:)]) {
+            [strongInternalDelegate commissioning:self failedWithError:error forDeviceID:commissioningID metrics:metrics];
+        } else if ([strongDelegate respondsToSelector:@selector(commissioning:failedWithError:metrics:)]) {
+            [strongDelegate commissioning:self failedWithError:error metrics:metrics];
+        }
+
+        // No matter what, notify our controller that we are done.  This is not
+        // ideal, because if the delegate queue is suspended we will never clear
+        // our state on the MTRDeviceController.  But we want this part to come
+        // after the delegate notification when the delegate is the controller
+        // itself.  Try to figure out a better way to do this.
+        [strongController commissioningDone:self];
```

Comment:
If you place this call after the dispatch_async, and the controller itself dispatches to its own queue, wouldn't this still ensure this work happens after the delegate callbacks are performed (i.e. two dispatch_async to the same queue)